### ISSUE TITLE
`esbuild@0.15.5` uses `compilerOptions.baseUrl` when resolving transitive dependencies

### ISSUE THREAD

**Issue by merceyz** [evanw/esbuild#2481]
**Describe the bug**

When resolving transitive dependencies `esbuild@0.15.5` tries to resolve them from `compilerOptions.baseUrl`.

Ref https://github.com/yarnpkg/berry/pull/4732
Ref https://github.com/evanw/esbuild/issues/2473#issuecomment-1217106596

**To Reproduce**

```sh
docker run --rm -it node:16.16.0 bash
cd $(mktemp -d)
yarn init -2
yarn set version 3.2.2
yarn config set nodeLinker node-modules
yarn add esbuild@0.15.5 react@18.2.0 react-dom@18.2.0
printf '{"compilerOptions": {"baseUrl": "."}}' > tsconfig.json
printf "import 'react-dom'" > index.ts
printf "this is invalid" > react.js
yarn esbuild --bundle ./index.ts
```

**Comment by evanw** [comment]
FWIW this is not specific to Yarn. This happens when npm is used instead of Yarn. It looks like this has been esbuild's behavior for quite a while.

**Comment by merceyz** [comment]
Indeed, that's why the reproduction uses the `node-modules` linker (`node_modules folder)

**Comment by evanw** [comment]
I'm still not totally sure what to do about this. One reason why TypeScript doesn't do this is that it doesn't check JavaScript files. Another reason is that (I think?) it doesn't type packages, which is basically the equivalent of esbuild's `--packages=external` setting, which would also work here. So I'm not sure if esbuild's behavior should be that path mappings in tsconfig files shouldn't apply to JavaScript files, or that they shouldn't apply to `node_modules` files, or something else.
